src/NUnitFramework/framework/Internal/Execution/WorkItem.cs

```diff
@@ -300,9 +315,35 @@ public virtual void Cancel(bool force)
 #endif
         }
 
+        #endregion
+
+        #region IDisposable Implementation
+
+        /// <summary>
+        /// Standard Dispose
+        /// </summary>
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
```

Comment:
I'm sorry; it crossed my mind if I should mention this originally because MSDN is out of date and misleads people. Real quick: since .NET 4.0 introduced the SafeHandle class, it's now an anti-pattern to use the `Dispose(bool)` pattern or to involve the finalizer. (The SafeHandle contains the finalizer logic and gets other special treatments. Any unmanaged resources (bare IntPtr handles) should go inside a SafeHandle class instead, and the SafeHandle, being itself a managed resource just gets disposed like any other managed object.)

Now I know we run on net35 and net20, but the same logic applies: the `WorkItem` class itself, and any subclass, will not and should not ever directly own bare unmanaged (IntPtr) handles. So as a safeguard we can remove everything except `public void Dispose() { _completionEvent.Dispose(); }`. (And we don't need the null check because the variable is readonly.)

Reply:
ManualResetEvent on net35 and net20 already contains the finalizer logic too, so it would be redundant for the `WorkItem` class to have finalizer logic. It's already guaranteed to be disposed without `WorkItem` doing anything special.

Reply:
I know that this is counter to a lot of common practice, so it's understandable that you might want confirmation that I'm not crazy on this. 😀 This comes up a lot at dotnet/csharplang and I can loop in sharwell or someone.

Reply:
@jnm2 I won't be changing this right away - it's also an anti-pattern to keep changing minor bits in code that is not yet even doing what it's supposed to do!!! I'll hold it in mind as I work on the next commit with actual functionality.

Based on what you are now saying, it sounds as if the initial code was correct and that WorkItem does not need to implement dispose. I understood that MRE basically took care of itself in different ways on each platform, so we didn't have to. I'm not in favor of implementing an empty Dispose "just in case". Or are you saying we need to implement and invoke DIspose solely for the purpose of getting the release done early?

Please give it some thought and answer at your leisure.

Reply:
> it's also an anti-pattern to keep changing minor bits in code that is not yet even doing what it's supposed to do!!!

Yes! :D

> Or are you saying we need to implement and invoke DIspose solely for the purpose of getting the release done early?

Yes, exactly. Deterministic release is always preferable to waiting for the garbage collector to finalize an object. One reason is that the garbage collector reacts to memory pressure but not to leaking handles so you can go for an arbitrarily large amount of time without a GC and yet still leaking kernel handles.

The finalization handling is a backstop, not a great one but a necessary one, and since it's built into ManualResetEvent we don't need to build it ourselves. But we'd prefer not to use the backstop.

Reply:
Do we believe this to be true on all platforms, including Mono?

Reply:
I checked Mono and it's using a SafeHandle. (https://github.com/mono/mono/blob/master/mcs/class/referencesource/mscorlib/system/threading/eventwaithandle.cs)

If it's not true in a platform's BCL, it's a bug in the platform just as much as it would be a bug in NUnit if we used a bare IntPtr handle without encapsulating it in a SafeHandle class.

.NET Framework and .NET Core are identical. Mono has been incrementally replacing its BCL with .NET Core's BCL implementation. Xamarin and Unity use Mono's BCL.

Reply:
Sounds like the right way to go then. Ironic to be doing what was called bad practice for so long.

Reply:
This is over 3 years old, but may be relevant: http://www.mono-project.com/docs/advanced/safehandles/

Reply:
@CharliePoole I don't think it is an issue since this is the way we have been doing it, but were you planning on updating this based on the comments?